dora/core/views.py

```diff
@@ -37,3 +48,105 @@ def ping(request):
 def trigger_error(request):
     division_by_zero = 1 / 0
     print(division_by_zero)
+
+
+@api_view()
+@permission_classes([permissions.AllowAny])
+def get_inclusion_connect_login_info(request):
+    # https://security.stackexchange.com/questions/147529/openid-connect-nonce-replay-attack
+    # https://stackoverflow.com/questions/46844285/difference-between-oauth-2-0-state-and-openid-nonce-parameter-why-state-cou
+    # https://blogs.aaddevsup.xyz/2019/11/state-parameter-in-mvc-application/
+    # https://stackoverflow.com/questions/35165793/what-attack-does-the-state-parameter-in-openid-connect-server-flow-prevent
+    # https://stackoverflow.com/questions/53246830/how-nonce-and-state-parameters-are-stored-and-transmitted-in-identityserver4
+    # https://stackoverflow.com/questions/46844285/difference-between-oauth-2-0-state-and-openid-nonce-parameter-why-state-cou/46859861#46859861
+    state = get_random_string(32)
+    nonce = get_random_string(32)
+    # Not sure hashing gets us anything…
+    hashed_nonce = make_password(nonce)
+
+    cache.set(
+        f"oidc-state-{state}",
+        {
+            "state": state,
+            "nonce": nonce,
+        },
+    )
+    return Response(
+        {
+            "url": f"{settings.IC_BASE_URL}auth?response_type=code&from=dora&client_id={settings.IC_CLIENT_ID}&scope=openid profile email&nonce={hashed_nonce}",
+            "state": state,
+        }
+    )
+
+
+@api_view(["POST"])
+@permission_classes([permissions.AllowAny])
+def get_inclusion_connect_user_info(request):
+    code = request.data.get("code")
+    state = request.data.get("state")
+    redirect_uri = request.data.get("redirect_uri")
+
+    stored_state = cache.get(f"oidc-state-{state}")
+    assert stored_state["state"] == state
+    stored_nonce = stored_state["nonce"]
+
+    try:
+        response = requests.post(
+            url=f"{settings.IC_BASE_URL}token",
+            headers={
+                "Content-Type": "application/x-www-form-urlencoded",
+            },
+            data={
+                "client_id": settings.IC_CLIENT_ID,
+                "client_secret": settings.IC_CLIENT_SECRET,
+                "code": code,
+                "grant_type": "authorization_code",
+                "redirect_uri": redirect_uri,
+            },
+        )
+        result = json.loads(response.content)
+
+        id_token = result["id_token"]
+
+        # TODO: décoder / verifier la signature
```

Comment:
En fait, compte-tenu du fait qu'ici l'ID token vient d'être récupéré via un appel direct du back au provider OIDC (et non par exemple, récupéré par le front puis transmis au back), la validation du token ne protège pas grand-chose : si le provider OIDC a été compromis, ou ses réponses sont modifiées par un attaquant, alors celui-ci pourra vraisemblablement aussi modifier la réponse de l'endpoint JWKS (ex. https://recette.connect.inclusion.beta.gouv.fr/realms/local/protocol/openid-connect/certs) et y injecter sa propre clé publique, et donc on aura beau vérifier la signature, elle sera valide même pour un token forgé.

Ce cas précis est d'ailleurs mentionné dans le document de l'IETF "OAuth 2.0 for Browser-Based Apps" : https://datatracker.ietf.org/doc/html/draft-ietf-oauth-browser-based-apps#section-9.8.3

> Performing OpenID Connect using the Authorization Code flow provides the benefit of the client not needing to verify the JWT signature, as the ID token will have been fetched over an HTTPS connection directly from the authorization server.